components/invitation_modal/invitation_modal.tsx

```diff
@@ -0,0 +1,396 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {Modal} from 'react-bootstrap';
+
+import {injectIntl, IntlShape} from 'react-intl';
+
+import {InviteToTeamTreatments} from 'mattermost-redux/constants/config';
+import deepFreeze from 'mattermost-redux/utils/deep_freeze';
+import {debounce} from 'mattermost-redux/actions/helpers';
+import {ActionFunc} from 'mattermost-redux/types/actions';
+import {SubscriptionStats} from 'mattermost-redux/types/cloud';
+
+import {Team} from 'mattermost-redux/types/teams';
+
+import {Channel} from 'mattermost-redux/types/channels';
+import {UserProfile} from 'mattermost-redux/types/users';
+
+import {trackEvent} from 'actions/telemetry_actions';
+
+import {isEmail} from 'mattermost-redux/utils/helpers';
+
+import ResultView, {ResultState, defaultResultState, InviteResults} from './result_view';
+import InviteView, {InviteState, defaultInviteState} from './invite_view';
+import NoPermissionsView from './no_permissions_view';
+import {As} from './invite_as';
+
+import './invitation_modal.scss';
+
+export type Props = {
+    show: boolean;
+    inviteToTeamTreatment: InviteToTeamTreatments;
+    actions: {
+        closeModal: () => void;
+        searchChannels: (teamId: string, term: string) => ActionFunc;
+        regenerateTeamInviteId: (teamId: string) => void;
+
+        searchProfiles: (term: string, options?: Record<string, string>) => Promise<{data: UserProfile[]}>;
+        sendGuestsInvites: (
+            currentTeamId: string,
+            channels: Channel[],
+            users: UserProfile[],
+            emails: string[],
+            message: string,
+        ) => Promise<{data: InviteResults}>;
+        sendMembersInvites: (
+            teamId: string,
+            users: UserProfile[],
+            emails: string[]
+        ) => Promise<{data: InviteResults}>;
+    };
+    currentTeam: Team;
+    currentChannelName: string;
+    invitableChannels: Channel[];
+    emailInvitationsEnabled: boolean;
+    isAdmin: boolean;
+    isCloud: boolean;
+    subscriptionStats?: SubscriptionStats | null;
+    cloudUserLimit: string;
+    canAddUsers: boolean;
+    canInviteGuests: boolean;
+    intl: IntlShape;
+}
+
+export const View = {
+    INVITE: 'INVITE',
+    RESULT: 'RESULT',
+} as const;
```

Comment:
just curious, why the `as const`?

Reply:
As a way to narrow the types from being strings to being only the strings listed. Some more details:
https://rlee.dev/stop-misusing-typescript-string-enums
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions